gateway/service/src/main/java/org/eclipse/ditto/gateway/service/endpoints/actors/CheckPermissionsActor.java

```diff
@@ -0,0 +1,417 @@
+     *
+     * @param edgeCommandForwarder the actor responsible for forwarding commands to edge services.
+     * @param sender the actor reference to the original sender.
+     * @param defaultTimeout the default timeout for async operations.
+     */
+    private CheckPermissionsActor(final ActorRef edgeCommandForwarder, final ActorRef sender,
+            final Duration defaultTimeout) {
+        this.edgeCommandForwarder = edgeCommandForwarder;
+        this.sender = sender;
+        this.defaultAskTimeout = defaultTimeout;
+    }
+
+    /**
+     * Creates {@link Props} for the {@link CheckPermissionsActor}.
+     *
+     * @param edgeCommandForwarder the actor responsible for forwarding commands to edge services.
+     * @param sender the actor reference to the original sender.
+     * @param defaultTimeout the default timeout for async operations.
+     * @return a {@link Props} object to create the actor.
+     */
+    public static Props props(final ActorRef edgeCommandForwarder, final ActorRef sender,
+            final Duration defaultTimeout) {
+        return Props.create(CheckPermissionsActor.class, edgeCommandForwarder, sender, defaultTimeout);
+    }
+
+    @Override
+    public Receive createReceive() {
+        return ReceiveBuilder.create()
+                .match(CheckPermissions.class, this::handleCheckPermissions)
+                .matchAny(msg -> logger.warn("Unknown message: <{}>", msg))
+                .build();
+    }
+
+    /**
+     * Handles the {@link CheckPermissions} message by initializing the permission results,
+     * grouping permission checks by entity, and performing policy retrievals.
+     *
+     * @param command the {@link CheckPermissions} message to handle.
+     */
+    private void handleCheckPermissions(final CheckPermissions command) {
+        final Map<String, Boolean> permissionResults = initializePermissionResults(command);
+        final Map<String, Map<String, PermissionCheckWrapper>> groupedByEntityId =
+                groupByEntityAndPolicyResource(command);
+
+        CompletableFuture<Void> allPolicyRetrievals = CompletableFuture.allOf(
+                groupedByEntityId.keySet().stream()
+                        .map(entityId -> retrieveOrSetPolicyId(entityId, groupedByEntityId.get(entityId), command))
+                        .toArray(CompletableFuture[]::new)
+        );
+
+        allPolicyRetrievals.thenRun(
+                        () -> handlePolicyRetrievalCompletion(command, permissionResults, groupedByEntityId))
+                .exceptionally(ex -> handleFailure(ex, command));
+    }
+
+    /**
+     * Initializes the permission results map with 'false' for each permission check.
+     *
+     * @param command the {@link CheckPermissions} command containing the permission checks.
+     * @return a map with the permission check results, initialized to {@code false}.
+     */
+    private Map<String, Boolean> initializePermissionResults(final CheckPermissions command) {
```

Comment:
I would prefer Java simple types, so instead of `Boolean`, please use `boolean` in the map.

Reply:
but Java Maps only support boxed types

Reply:
Ooops, absolutely right 👍